### ISSUE TITLE
Creation of session should return `F[Session[F]]`

### ISSUE THREAD

**Issue by gvolpe** [neotypes/neotypes#38]
Creation of mutable state should be wrapped in `F` for the reasons stated [here](https://typelevel.org/blog/2018/06/07/shared-state-in-fp.html) and [here](https://www.reddit.com/r/scala/comments/be6qxp/stm_software_transactional_memory_examples_in/). Fabio Labella also gave a [great talk](https://vimeo.com/294736344) on the topic.

```scala
val session: Session[IO] = driver.session().asScala[IO] // this can leak

val makeSession: IO[Session[IO]] = driver.session.asScala[IO] // this is the action of creating a session

makeSession.flatMap { s => // here we create a region of sharinn
  useSession(s) // whoever needs access to the session should take it as a parameter 
 // in this way the sharing of the session is under our control in a narrow scope (inside `flatMap`)
}
```

Right now I get around by just wrapping its creation in `IO` but ideally the library should do this.

This probably doesn't make sense when using `Future` since it breaks referential transparency but for the rest of us using FP libraries this is important.

Let me know if I can be of any help.
```

**Comment by BalmungSan** [comment]
Could you please expand why it should be wrapped in the effect itself?
I was planning in adding a `session` method to the **Driver**, to create a **Session** directly.
Thus, that could be a good place to add that.

However, I still do not understand, what is shared about the session?
I believe it is kind of immutable. However, I really do not know how the internal Neo4j session is implemented just the wrapper in this library.

**Comment by gvolpe** [comment]
@BalmungSan have you read the articles I shared or just watched the talk I linked? It's all there :)

**Comment by BalmungSan** [comment]
@gvolpe well yes, but still I do not see the reason for the library to do that.
Usually, I will have a class that encapsulates all **Neo4j** operations, I will just pass the unique instance to my services as the `HttpClient` example.
In that class I will receive the `Session` object in the constructor, and I will make that constructor private.
And the companion object will have a factory returning the instance of my class wrapped on a `Resource[F, C]` which will close the session after use.

**Comment by gvolpe** [comment]
> well yes, but still I do not see the reason for the library to do that.

Creation of mutation *is a side-effect*. This should be supported by the library unless having mutability and side-effects is what we want to promote.

What you're doing is fine and is what I have to do as well. But I don't agree this should be let off to the user.

---

If you're looking for examples look at the creation of `Ref`, `Deferred`, etc in `cats-effect`. Same for the equivalent data structures in `zio`. Their creation is wrapped in `F` because creation of mutable state is a side-effect.

Totally recommend watching Fabio's talk on the topic.

Gotta go now but can give more examples later if you remain unconvinced.

**Comment by BalmungSan** [comment]
Sorry, I did not make myself clear.
I do kwon that creation of mutable state should be wrapped in an `F`.
What I do not know, is if a **Session** is mutable state, the wrapper is not, but probably the underlying **Neo4j** session is.

@dimafeng What do you think, I would be in favor of returning `F[Driver[F]]` and having a method in the **Driver** as the only way to create a `Session[F]`.

**Comment by gvolpe** [comment]
> What I do not know, is if a Session is mutable state

It is mutable state indeed. Here's an example that leaks:

```scala
object Main {
   val s = driver.session().asScala[IO]
   
   val program =
     ServiceOne.foo(s) *> ServiceTwo.foo(s)

   /// more services and more logic

   // access to session where it shouldn't be accessed. This is more obvious in larger programs.
   s.beginTransaction.flatMap { tx =>
     c"create (p:Record {personId: 123)".query[Unit].execute(tx)
   }

}

object ServiceOne {
  def foo(session: Session[IO]): IO[Unit] = ??? // do something with the session
}

object ServiceTwo {
  def foo(session: Session[IO]): IO[Unit] = ??? // do something with the session
}

```

In order to control the region of sharing we use `flatMap` and explicitly share the mutable state as arguments of other functions.

```scala
object Main {
   val mkSession: IO[Session[IO]] = IO(driver.session().asScala[IO])
   
   val program =
     mkSession.flatMap { s => // specific region of sharing
       ServiceOne.foo(s) *> ServiceTwo.foo(s)
     }

   /// more services and more logic

   // the same sessions cannot be accessed here now
   mkSession.flatMap { s => // this is a different session
     s.beginTransaction.flatMap { tx =>
       c"create (p:Record {personId: 123)".query[Unit].execute(tx)
     }
   }

}

object ServiceOne {
  def foo(session: Session[IO]): IO[Unit] = ??? // do something with the session
}

object ServiceTwo {
  def foo(session: Session[IO]): IO[Unit] = ??? // do something with the session
}
```

---

> Given a F[Driver[F]], creating a Session from such Driver should also be wrapped in a F, right? Thus, one would need to flatMap it.

That's correct. If we were to do this just for principled FP libraries this would be a resource. i.e. in `cats-effect` it should be `Resource[F, Driver[F]]` and that's effectively how I use it.

> The Driver is probably a shared state too

Answered above.

**Comment by BalmungSan** [comment]
Just out of curiosity _(as I agree with you, but I fail to see it)_ what exactly is wrong with the second use of the session?
It will just return an `IO`, so it wont be executed _(unless an unsafe call is done)_. Thus, either one would need to chain it with program using `flatMap` or either of them will just be discarded.
Probably I am missing something obvious.

**Comment by gvolpe** [comment]
No worries. The concept of mutable state sharing is so simple that is hard to get at first :)

In the first example (the one that leaks a session) we could potentially have *unintended access* to our session. This compares pretty much to global state. The variable `s` can be accessed anywhere in `Main` but it could also be accessed from anywhere in the program by doing `Main.s` (yes, you could restrict this by adding `private` but the first problem remains).

In the second (and safe) example we are making explicit when some mutable state is being created as we don't want to unintentionally share it. So it is wrapped in `IO` (or any other arbitrary effect). If we call `mkSession` is obvious that we are creating a new session, you only need to look at its type signature. And if we want to share this session then you need to create an explicit *region of sharing* which is dictated by `flatMap` and pass it as an argument.

Again, I recommend watching Fabio's talk, he does a great job explaining this.

**Comment by dimafeng** [comment]
@gvolpe thank you for raising this question. I do see your point. Do I understand correctly that the proposed change is to wrap session creation in an effect inside `asScala` method?

I agree that it looks a bit redundant with scala `Future` but I think that won't be an issue.

**Comment by BalmungSan** [comment]
@dimafeng Just FYI I have been working in this proposal, going to open a WIP PR in a few minutes inc ase you can / want to check it.

**Comment by dimafeng** [comment]
@BalmungSan no worries. I was trying to understand the proposed change.

**Comment by gvolpe** [comment]
@dimafeng the implementation is up to you. My proposal is for changing the type signature, which means the implementation should be changed as well but doesn't specify how as long as it behaves as expected.

**Comment by sloshy** [comment]
I'd like to go a step further and actually suggest dropping `Future` support completely.

I don't necessarily agree that returning an `F[Driver[F]]` in all cases is the best decision. For cats-effect and similar, it makes sense since as @gvolpe mentioned creating/closing it is a side-effect, but as mentioned `Future` has nothing to do with handling side-effects so what this does is it creates an inconsistency in the API. Now, for users of `Future`, the entirety of your neotypes-using program is wrapped in a `Future` that takes an implicit `ExecutionContext` on every map/flatmap and that just seems messy and easy to cause problems with.

To build on this, I think the driver - in all public APIs - should be a `cats.effect.Resource` with no way to create one otherwise. Yes I know that would mean rebasing the project on top of cats-effect for the base module, but I think this would provide us several benefits and reduce the maintenance burden, including:

* Getting rid of `neotypes.Async` for internal use as we can just use the cats-effect one internally
* Not needing to provide any `Future`-specific code - that's handled by `unsafeToFuture`, et al, and should arguably be faster than providing `Future` anyway since there is usually less context-shifting w/ IO, Task, etc.
* Being able to provide the driver as a proper `Resource` that, in all cases, is properly used (whether ran as `Future` or not)
* Free Monix and ZIO support since Monix 3 depends on with cats-effect already and ZIO has cats effect integration
* Potential to control `ExecutionContext` usage inside of transactions in a consistent, predictable way that doesn't change based on the type "powering" the library.

The TL;DR is that the library should not be providing any possible way for the user to mess up. The driver being wrapped in a `Future` I think is only going to cause confusion and problems with every flatmap, not to mention it isn't idiomatic usage of `Future`. And the fact that users would have to manually make it into a `Resource` is just asking for trouble.

**Comment by gvolpe** [comment]
Definitely agree with all the points made by @sloshy .

Dropping support for `Future` is probably a drastic decision if there are many users out there. If that's not an option I think separating the API between pure / impure would be another option with the maintenance cost this will imply.

But I agree that instead of `F[Driver[F]]` it should be `Resource[F, Driver[F]]` and that's encouraged as well in #39 but in user's land instead of being natively supported by `neotypes`. 

What I'd do is to open an issue / raise a poll on Twitter and Reddit / post on Gitter to gain an understanding of how many users are using it with `Future` before making a decision.

**Comment by BalmungSan** [comment]
I have to agree that using `cats.effect.Async` instead of our own `Async` _(as well as using `Resource` and the other proposed changes)_ would be better for _"pure"_ users like us. Especially because that was the main purpose of **cats-effect**, to serve as an abstraction for library authors which allowed users to use their preferred `IO` monad.

However, we have to agree that in a sense that goes against the original principles of the library like:

> The project aims to provide seamless integration with most popular scala infrastructures such as **lightbend (Akka, Akka-http, Lagom, etc)**, typelevel (cats, http4s, etc), twitter (finch, etc)...

Also, I do not see how we could maintain two _"versions"_ of the library without one causing problems with the other, they would have to be two different libraries.
Or maybe, having an impure core and a pure wrapper on top of it? - But I am not sure if that would really provide the benefits we want.

Finally, while I definitively agree with doing some kind of poll, unfortunately that would not include the opinion of `Future` _(bad joke intended)_ users which may not agree with the decision we made now.

**Comment by dimafeng** [comment]
I agree and disagree with @sloshy's  points 😁 
1. As @BalmungSan mentioned, the original goal of the library was `to provide seamless integration with most popular scala infrastructures` + `the driver depends on shapeless and neo4j Java driver.`. Dropping direct `Future` support introduces extra complexity for `Future` users, and most importantly (I guess), the dependency on `cats` scares `Future` users (an example of it is `doobie` -- it's not well adopted by Akka/Play Framework users)'
2. I fully agree with `Future[Driver[Future]]` being really painful to use and it's definitely not idiomatic.
3. I still believe in the ability to satisfy both groups. One of the options is to use implicit classes, e.g.:
```
GraphDatabase[Future](....).driver() // Driver[Future]

GraphDatabase[IO](....).driver() // IO[Driver[IO]]
```
And same approach for `Session` creation (potentially returning `Resource[F, Session[F]]` for effects and `Session[F]` for future). @gvolpe, Am I describing your option `separating the API between pure / impure`? 😁 

There are a couple of other options using naming convention like `driver` and `driverUnsafe` but I don't like this approach

**Comment by BalmungSan** [comment]
I think some kind of `Resource` **typeclass** from our side may work.
Since  it seems the only area of friction is really just the **Driver** / **Session** creation and lifecycle management. For the rest of the code, it seems both worlds can coexist without problems.

Maybe something like:

```scala
trait Resource[F[_]] {
  type R[_, _] // I think I have a syntax error here, since the first parameter should like F[_].

  def create[T](t: => T): R[F, T]
}
```

Where for **IO** `R[_, _] == Resource[IO, T]` and for **Future** `R[_, _] = Id[Id, T]`
Thus `Future` users could do this:

```
val driver: Driver[Future] = GraphDatabase[Future].driver(...)
val session: Session[Future] = driver.session(...)
```

While **IO** _(and friend)_ users will have:

```
val session: Resource[IO, Session[IO]] = for {
  driver <- GraphDatabase[IO].driver(...)
  session <- driver.session(...)
} yield session
```

_(For **Zio** folks it will be a `ZManaged` instead)_.

This is just an sketch I just made on top of my head, I am not totally sure if both cases can be supported, but I guess it is worth exploring.

**Comment by gvolpe** [comment]
> @gvolpe, Am I describing your option separating the API between pure / impure? :grin: 

Seems like it :smile:

**Comment by BalmungSan** [comment]
@dimafeng @gvolpe @sloshy 

I just made a **PoC** of the my previous idea, and it works like a charm _(I think 😅)_.
[Here](https://scastie.scala-lang.org/BalmungSan/YYfJ04LeSr2gTTfBpggkaw/35) is the code, I tested it with Scala `2.12` & `2.11`. - Just one thing, and it is that the **IO** example requires `-Ypartial-unification` on `2.12` and [**sbt-partial-unification**](https://github.com/fiadliel/sbt-partial-unification) on `2.11`, however I doubt anyone using **cats** will not have these already, and I believe it is not necessary for **Future**.

The example provides the desired semantics for both parties:
+ Plain & eager `Driver[Future]`.
+ Contained & lazy `Resource[IO, Driver[IO]]`.

The only thing I would like to have would be making `close` private... But, **Future** users need it exposed.

Anyways, I would like to know your opinions about this.
If we all agree, I will include this in #39

**Comment by dimafeng** [comment]
@BalmungSan looks good to me!

**Comment by dimafeng** [comment]
Fixed/implemented in the scope of https://github.com/neotypes/neotypes/pull/39
